### PR TITLE
Experimental jupyterhub support - DO NOT MERGE

### PR DESCRIPTION
**Pull request by adamstruck**
To support spawner jupyter notebook servers from jupyterhub the following changes were made:

* added metadata bucket to boltdb
* added metadata field to UpdateStatusRequest message
* DockerCmds name the container JobIDStepNum to allow inspection
    * could also be used to kill the container?
* Port bindings in DockerCmd hardcoded for example
    * unless fixed only one container can be run at a time... 
* runStep returns async process
    *  allows stdout polling (steaming) while job is running
* container metadata (from docker inspect) is dumped into the metadata bucket in boltdb in the form <JobIDStepNum>: <json string>

I don't intended on this PR ever being merged - just wanted to create an area where we could discuss the changes I made.

---

- [X] `make` does not report any errors
- [ ] `make tidy` does not report any errors (except for package name related errors)
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are documentation for these changes OR
- [ ] These changes do not require documentation because _____

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

### CODE REVIEW

File: src/tes/worker/engine.go [hunk 1]
```diff
@@ -152,19 +156,72 @@ func (eng *engine) runJob(ctx context.Context, sched *scheduler.Client, jobR *pb
...
-		if stepLog != nil {
-			// Send the scheduler service a job status update
-			statusReq := &pbr.UpdateStatusRequest{
-				Id:   job.JobID,
-				Step: int64(stepNum),
-				Log:  stepLog,
-			}
-			sched.UpdateJobStatus(ctx, statusReq)
+		// Start task step asynchronously
+		dcmd, async_proc, _ := eng.runStep(mapper, step, fmt.Sprintf("%v-%v", job.JobID, int64(stepNum)))
```
Comment (buchanae) [inline src/tes/worker/engine.go]: stepID := fmt.Sprintf(...) would be easier to read.

File: src/tes/worker/engine.go [hunk 2]
```diff
@@ -241,15 +298,60 @@ func (eng *engine) downloadInputs(mapper *FileMapper, store *storage.Storage) er
...
+func (eng *engine) queryContainerMetadata(containerName string, retries int) (string, error) {
+	// TODO is this the right approach?
+	if retries >= 10 {
+		return "", fmt.Errorf("Error getting metadata for container: %s", containerName)
+	}
+	
+	metadata_query_args := []string{
+		"inspect", 
+		containerName,
+	}
```
Comment (buchanae) [inline src/tes/worker/engine.go]: Unless you need to add more arguments in a loop, you probably don't need this.

File: src/tes/worker/engine.go [hunk 3]
```diff
@@ -241,15 +298,60 @@ func (eng *engine) downloadInputs(mapper *FileMapper, store *storage.Storage) er
 	return nil
 }
 
+func (eng *engine) queryContainerMetadata(containerName string, retries int) (string, error) {
+	// TODO is this the right approach?
```
Comment (buchanae) [inline src/tes/worker/engine.go]: There's a more idiomatic Go approach that doesn't use recursion, and instead uses a for loop, a goroutine, and a done channel. I suggest we do that. It looks something like...

```
done := make(chan struct{})
result := ""
go func() {
  for i := 0; i < maxRetries; i++ {
    // ...run docker inspect
    if inspectWorked{
      result = inspectOutput
      close(done)
      return
    }
  }
}()
<-done

// parse inspect results
```

File: src/tes/worker/engine.go [hunk 4]
```diff
@@ -280,19 +382,43 @@ func (eng *engine) runStep(mapper *FileMapper, step *pbe.DockerExecutor) (*pbe.J
...
+	if dcmd.Stderr != nil {
+		stderrText = readFileHead(dcmd.Stderr.Name())
+	}
+
+	steplog := &pbe.JobLog{
+		Stdout:   stdoutText,
+		Stderr:   stderrText,
+	}
+
+	return steplog
```
Comment (buchanae) [inline src/tes/worker/engine.go]: Duplicate code in finalizeStep()

File: src/tes/worker/engine.go (hunk 1 above)
Comment (buchanae) [inline src/tes/worker/engine.go]: I don't think you need int64()

File: src/tes/worker/engine.go [hunk 5]
```diff
@@ -152,19 +156,72 @@ func (eng *engine) runJob(ctx context.Context, sched *scheduler.Client, jobR *pb
...
+			Factor: 2,
+			Jitter: false,
+		}
+
+	  PollLoop:
+		for {
+			select {
+			case cmd_err := <-done:
+				stepLog := eng.finalizeStep(dcmd, cmd_err)
+				// Send the scheduler service a final job status update that includes
```
Comment (buchanae) [inline src/tes/worker/engine.go]: At first I was confused by this comment, but I later realized that finalizeStep() actually sets the error code. We can find a way to make this more clear.

**Comment by buchanae** [comment]
Oh, and there's a `make tidy` command which will automate formatting of the code, by the way.

**Comment by adamstruck** [comment]
* implemented cancel job
* all stdout / stderr is streamed into the logs
    * not sure that this is the behavior we want
    * boltdb only is storing the head of the stdout and stderr files

**Comment by adamstruck** [comment]
Split into:

https://github.com/bmeg/task-execution-server/pull/56
https://github.com/bmeg/task-execution-server/pull/57

Networking changes to schema: bmeg/task-execution-schemas#4

**Pull request closed without merging**